File: src/fast_resume/adapters/claude.py
"""Claude Code session adapter."""

import orjson
from datetime import datetime
from pathlib import Path

from ..config import AGENTS, CLAUDE_DIR
from ..logging_config import log_parse_error
from .base import BaseSessionAdapter, ErrorCallback, ParseError, Session, truncate_title


class ClaudeAdapter(BaseSessionAdapter):
    """Adapter for Claude Code sessions."""

    name = "claude"
    color = AGENTS["claude"]["color"]
    badge = AGENTS["claude"]["badge"]
    supports_yolo = True

    def __init__(self, sessions_dir: Path | None = None) -> None:
        self._sessions_dir = sessions_dir if sessions_dir is not None else CLAUDE_DIR
        self._project_indexes: dict[
            Path, tuple[float, dict[str, tuple[str, float]]]
        ] = {}

    def find_sessions(self) -> list[Session]:
        """Find all Claude Code sessions."""
        if not self.is_available():
            return []

        sessions = []
        for project_dir in self._sessions_dir.iterdir():
            if not project_dir.is_dir():
                continue

            for session_file in project_dir.glob("*.jsonl"):
                # Skip agent subprocesses
                if session_file.name.startswith("agent-"):
                    continue

                session = self._parse_session_file(session_file)
                if session:
                    sessions.append(session)

        return sessions

    def _parse_session_file(
        self, session_file: Path, on_error: ErrorCallback = None
    ) -> Session | None:
        """Parse a Claude Code session file."""
        try:
            first_user_message = ""
            ai_title = ""
            directory = ""
            timestamp = datetime.fromtimestamp(session_file.stat().st_mtime)
            messages: list[str] = []
            # Count conversation turns (user + assistant, not tool results)
            turn_count = 0

            with open(session_file, "rb") as f:
                for line in f:
                    if not line.strip():
                        continue
                    try:
                        data = orjson.loads(line)
                    except orjson.JSONDecodeError:
                        # Skip malformed lines within the file
                        continue

                    msg_type = data.get("type", "")

                    # Claude Code's auto-generated session title (shown in its
                    # own Resume UI). It gets rewritten as the session evolves,
                    # so keep the latest non-empty one.
                    if msg_type == "ai-title" and data.get("aiTitle"):
                        ai_title = data["aiTitle"]

                    # Get directory from user message
                    if msg_type == "user" and not directory:
                        directory = data.get("cwd", "")

                    # Process user messages
                    if msg_type == "user":
                        msg = data.get("message", {})
                        content = msg.get("content", "")

                        # Check if this is a real human input or automatic tool result
                        is_human_input = False
                        if isinstance(content, str):
                            is_human_input = True
                            if not data.get("isMeta") and not content.startswith(
                                ("<command", "<local-command")
                            ):
                                messages.append(f"» {content}")
                                if not first_user_message and len(content) > 10:
                                    first_user_message = content
                        elif isinstance(content, list):
                            # Check first part - if it's text (not tool_result), it's human
                            first_part = content[0] if content else {}
                            if isinstance(first_part, dict):
                                part_type = first_part.get("type", "")
                                if part_type == "text":
                                    is_human_input = True
                                # tool_result means automatic response, not human input

                            for part in content:
                                if (
                                    isinstance(part, dict)
                                    and part.get("type") == "text"
                                ):
                                    text = part.get("text", "")
                                    messages.append(f"» {text}")
                                    if not first_user_message:
                                        first_user_message = text
                                elif isinstance(part, str):
                                    messages.append(f"» {part}")

                        if is_human_input:
                            turn_count += 1

                    # Extract assistant content
                    if msg_type == "assistant":
                        msg = data.get("message", {})
                        content = msg.get("content", "")
                        has_text = False
                        if isinstance(content, str) and content:
                            messages.append(f"  {content}")
                            has_text = True
                        elif isinstance(content, list):
                            for part in content:
                                if (
                                    isinstance(part, dict)
                                    and part.get("type") == "text"
                                ):
                                    text = part.get("text", "")
                                    if text:
                                        messages.append(f"  {text}")
                                        has_text = True
                                elif isinstance(part, str):
                                    messages.append(f"  {part}")
                                    has_text = True
                        if has_text:
                            turn_count += 1

            # Skip sessions with no actual user message
            if not first_user_message:
                return None

            # Prefer Claude's own list title when available: a /rename display
            # name (sessions-index.json), then Claude's auto-generated aiTitle
            # (the title shown in its Resume UI), then the first user message.
            title_source = (
                self._get_index_title(session_file) or ai_title or first_user_message
            )
            title = truncate_title(title_source)

            # Skip sessions with no actual conversation content
            if not messages:
                return None

            full_content = "\n\n".join(messages)

            return Session(
                id=session_file.stem,
                agent=self.name,
                title=title,
                directory=directory,
                timestamp=timestamp,
                content=full_content,
                message_count=turn_count,
            )
        except OSError as e:
            error = ParseError(
                agent=self.name,
                file_path=str(session_file),
                error_type="OSError",
                message=str(e),
            )
            log_parse_error(
                error.agent, error.file_path, error.error_type, error.message
            )
            if on_error:
                on_error(error)
            return None
        except (KeyError, TypeError, AttributeError) as e:
            error = ParseError(
                agent=self.name,
                file_path=str(session_file),
                error_type=type(e).__name__,
                message=str(e),
            )
            log_parse_error(
                error.agent, error.file_path, error.error_type, error.message
            )
            if on_error:
                on_error(error)
            return None

    def _get_project_index_mtime(self, project_dir: Path) -> float:
        """Return the mtime of a Claude project sessions-index file."""
        try:
            return (project_dir / "sessions-index.json").stat().st_mtime
        except OSError:
            return 0.0

    def _parse_index_timestamp(self, value: str) -> float:
        """Parse a Claude sessions-index ISO timestamp to a unix timestamp."""
        if not value:
            return 0.0

        try:
            return datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp()
        except ValueError:
            return 0.0

    def _load_project_index(self, project_dir: Path) -> dict[str, tuple[str, float]]:
        """Load Claude session titles from a project's sessions-index.json."""
        index_mtime = self._get_project_index_mtime(project_dir)
        cached = self._project_indexes.get(project_dir)
        if cached and cached[0] == index_mtime:
            return cached[1]

        session_titles: dict[str, tuple[str, float]] = {}
        if index_mtime:
            try:
                with open(project_dir / "sessions-index.json", "rb") as f:
                    data = orjson.loads(f.read())

                entries = data.get("entries", [])
                if isinstance(entries, list):
                    for entry in entries:
                        if not isinstance(entry, dict):
                            continue

                        session_id = entry.get("sessionId", "")
                        summary = entry.get("summary", "")
                        if not isinstance(session_id, str) or not session_id:
                            continue
                        if not isinstance(summary, str) or not summary.strip():
                            continue

                        entry_mtime = index_mtime
                        modified = entry.get("modified", "")
                        if isinstance(modified, str):
                            entry_mtime = max(
                                entry_mtime, self._parse_index_timestamp(modified)
                            )

                        file_mtime = entry.get("fileMtime")
                        if isinstance(file_mtime, int | float):
                            entry_mtime = max(entry_mtime, file_mtime / 1000)

                        session_titles[session_id] = (
                            summary.strip(),
                            entry_mtime,
                        )
            except OSError, orjson.JSONDecodeError:
                session_titles = {}

        self._project_indexes[project_dir] = (index_mtime, session_titles)
        return session_titles

    def _get_index_title(self, session_file: Path) -> str:
        """Return Claude's indexed session title for a session file."""
        entry = self._load_project_index(session_file.parent).get(session_file.stem)
        return entry[0] if entry else ""

    def _scan_session_files(self) -> dict[str, tuple[Path, float]]:
        """Scan all Claude Code session files."""
        current_files: dict[str, tuple[Path, float]] = {}

        for project_dir in self._sessions_dir.iterdir():
            if not project_dir.is_dir():
                continue

            project_index = self._load_project_index(project_dir)

            for session_file in project_dir.glob("*.jsonl"):
                if session_file.name.startswith("agent-"):
                    continue

                try:
                    mtime = session_file.stat().st_mtime
                except OSError:
                    continue
                session_id = session_file.stem
                index_entry = project_index.get(session_id)
                if index_entry:
                    mtime = max(mtime, index_entry[1])
                current_files[session_id] = (session_file, mtime)

        return current_files

    def get_resume_command(self, session: Session, yolo: bool = False) -> list[str]:
        """Get command to resume a Claude Code session."""
        cmd = ["claude"]
        if yolo:
            cmd.append("--dangerously-skip-permissions")
        cmd.extend(["--resume", session.id])
        return cmd

File: src/adapters/claude.rs
use std::collections::HashMap;
use std::fs;
use std::io::{BufRead, BufReader};
use std::path::{Path, PathBuf};

use serde_json::Value;
use walkdir::WalkDir;

use crate::config;
use crate::model::{RawAdapterStats, Session, file_mtime_seconds, file_timestamp, truncate_title};

use super::shared::{
    content_texts, incremental_from_files, incremental_from_files_streaming,
    parse_timestamp_seconds, raw_stats_for_tree, string_at, text_from_part,
};
use super::{Adapter, IncrementalScan, KnownSessions, SessionCallback};

pub struct ClaudeAdapter {
    sessions_dir: PathBuf,
}

impl Default for ClaudeAdapter {
    fn default() -> Self {
        Self {
            sessions_dir: config::claude_dir(),
        }
    }
}

impl ClaudeAdapter {
    #[allow(dead_code)]
    pub fn new(sessions_dir: PathBuf) -> Self {
        Self { sessions_dir }
    }

    fn parse_session(&self, path: &Path) -> Option<Session> {
        let file = fs::File::open(path).ok()?;
        let mut directory = String::new();
        let mut first_user_message = String::new();
        let mut messages = Vec::new();
        let mut turns = 0usize;

        for line in BufReader::new(file).lines().map_while(Result::ok) {
            if line.trim().is_empty() {
                continue;
            }
            let Ok(data) = serde_json::from_str::<Value>(&line) else {
                continue;
            };
            let msg_type = data.get("type").and_then(Value::as_str).unwrap_or_default();

            if msg_type == "user" {
                if directory.is_empty() {
                    directory = string_at(&data, &["cwd"]);
                }
                let content = data
                    .pointer("/message/content")
                    .cloned()
                    .unwrap_or(Value::Null);
                let mut is_human_input = false;
                match content {
                    Value::String(text) => {
                        is_human_input = true;
                        let is_meta = data.get("isMeta").and_then(Value::as_bool).unwrap_or(false);
                        if !is_meta
                            && !text.starts_with("<command")
                            && !text.starts_with("<local-command")
                        {
                            messages.push(format!("» {text}"));
                            if first_user_message.is_empty() && text.chars().count() > 10 {
                                first_user_message = text;
                            }
                        }
                    }
                    Value::Array(parts) => {
                        if parts
                            .first()
                            .and_then(|part| part.get("type"))
                            .and_then(Value::as_str)
                            == Some("text")
                        {
                            is_human_input = true;
                        }
                        for part in parts {
                            if let Some(text) = text_from_part(&part) {
                                messages.push(format!("» {text}"));
                                if first_user_message.is_empty() {
                                    first_user_message = text;
                                }
                            } else if let Some(text) = part.as_str() {
                                messages.push(format!("» {text}"));
                            }
                        }
                    }
                    _ => {}
                }
                if is_human_input {
                    turns += 1;
                }
            } else if msg_type == "assistant" {
                let content = data
                    .pointer("/message/content")
                    .cloned()
                    .unwrap_or(Value::Null);
                let mut has_text = false;
                for text in content_texts(&content) {
                    messages.push(format!("  {text}"));
                    has_text = true;
                }
                if has_text {
                    turns += 1;
                }
            }
        }

        if first_user_message.is_empty() || messages.is_empty() {
            return None;
        }

        let title_source = claude_index_title(path).unwrap_or(first_user_message);
        let title = truncate_title(&title_source, 100, true);
        let mut session = Session::new(
            path.file_stem()?.to_string_lossy(),
            self.name(),
            title,
            directory,
            file_timestamp(path),
            messages.join("\n\n"),
            turns,
        );
        session.mtime = file_mtime_seconds(path);
        Some(session)
    }

    fn scan_session_files(&self) -> HashMap<String, (PathBuf, f64)> {
        let mut current_files = HashMap::new();
        let Ok(projects) = fs::read_dir(&self.sessions_dir) else {
            return current_files;
        };

        for project in projects.filter_map(Result::ok) {
            let project_dir = project.path();
            if !project_dir.is_dir() {
                continue;
            }
            let project_index = claude_project_index(&project_dir);
            let Ok(files) = fs::read_dir(&project_dir) else {
                continue;
            };
            for file in files.filter_map(Result::ok) {
                let path = file.path();
                if path.extension().and_then(|e| e.to_str()) != Some("jsonl") {
                    continue;
                }
                if path
                    .file_name()
                    .and_then(|name| name.to_str())
                    .is_some_and(|name| name.starts_with("agent-"))
                {
                    continue;
                }
                let Some(session_id) = path
                    .file_stem()
                    .and_then(|stem| stem.to_str())
                    .map(ToString::to_string)
                else {
                    continue;
                };
                let mut mtime = file_mtime_seconds(&path);
                if let Some((_, index_mtime)) = project_index.get(&session_id) {
                    mtime = mtime.max(*index_mtime);
                }
                current_files.insert(session_id, (path, mtime));
            }
        }

        current_files
    }
}

impl Adapter for ClaudeAdapter {
    fn name(&self) -> &'static str {
        "claude"
    }

    fn supports_yolo(&self) -> bool {
        true
    }

    fn find_sessions(&self) -> Vec<Session> {
        if !self.sessions_dir.exists() {
            return Vec::new();
        }
        let mut sessions = Vec::new();
        for entry in WalkDir::new(&self.sessions_dir)
            .min_depth(2)
            .max_depth(2)
            .into_iter()
            .filter_map(Result::ok)
        {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) != Some("jsonl") {
                continue;
            }
            if path
                .file_name()
                .and_then(|name| name.to_str())
                .is_some_and(|name| name.starts_with("agent-"))
            {
                continue;
            }
            if let Some(session) = self.parse_session(path) {
                sessions.push(session);
            }
        }
        sessions
    }

    fn find_sessions_incremental(&self, known: &KnownSessions) -> IncrementalScan {
        incremental_from_files(self.name(), known, self.scan_session_files(), |path| {
            self.parse_session(path)
        })
    }

    fn find_sessions_incremental_streaming(
        &self,
        known: &KnownSessions,
        on_session: &mut SessionCallback<'_>,
    ) -> IncrementalScan {
        incremental_from_files_streaming(
            self.name(),
            known,
            self.scan_session_files(),
            |path| self.parse_session(path),
            on_session,
        )
    }

    fn resume_command(&self, session: &Session, yolo: bool) -> Vec<String> {
        let mut cmd = vec!["claude".to_string()];
        if yolo {
            cmd.push("--dangerously-skip-permissions".to_string());
        }
        cmd.extend(["--resume".to_string(), session.id.clone()]);
        cmd
    }

    fn raw_stats(&self) -> RawAdapterStats {
        raw_stats_for_tree(self.name(), &self.sessions_dir, "jsonl")
    }
}

fn claude_index_title(session_file: &Path) -> Option<String> {
    let session_id = session_file.file_stem()?.to_string_lossy();
    claude_project_index(session_file.parent()?)
        .get(session_id.as_ref())
        .map(|(title, _)| title.clone())
}

fn claude_project_index(project_dir: &Path) -> HashMap<String, (String, f64)> {
    let mut titles = HashMap::new();
    let index_file = project_dir.join("sessions-index.json");
    let index_mtime = file_mtime_seconds(&index_file);
    let Ok(data) = serde_json::from_slice::<Value>(&fs::read(index_file).unwrap_or_default())
    else {
        return titles;
    };
    let Some(entries) = data.get("entries").and_then(Value::as_array) else {
        return titles;
    };
    for entry in entries {
        let session_id = string_at(entry, &["sessionId"]);
        let summary = string_at(entry, &["summary"]);
        if session_id.is_empty() || summary.trim().is_empty() {
            continue;
        }
        let modified = parse_timestamp_seconds(&string_at(entry, &["modified"])).unwrap_or(0.0);
        let file_mtime = entry
            .get("fileMtime")
            .and_then(Value::as_f64)
            .map(|value| value / 1000.0)
            .unwrap_or(0.0);
        let mtime = index_mtime.max(modified).max(file_mtime);
        titles.insert(session_id, (summary.trim().to_string(), mtime));
    }
    titles
}

#[cfg(test)]
mod tests {
    use std::fs;

    use serde_json::json;
    use tempfile::tempdir;

    use crate::adapters::Adapter;

    use super::*;

    #[test]
    fn uses_sessions_index_title() {
        let temp = tempdir().unwrap();
        let projects = temp.path().join("projects");
        let project = projects.join("project-a");
        fs::create_dir_all(&project).unwrap();

        fs::write(
            project.join("session-rename.jsonl"),
            [
                json!({
                    "type": "user",
                    "cwd": "/work/app",
                    "message": {"content": "Original first prompt for this session"}
                })
                .to_string(),
                json!({"type": "assistant", "message": {"content": "Response"}}).to_string(),
            ]
            .join("\n"),
        )
        .unwrap();
        fs::write(
            project.join("sessions-index.json"),
            json!({
                "version": 1,
                "entries": [{
                    "sessionId": "session-rename",
                    "summary": "Renamed Claude thread"
                }]
            })
            .to_string(),
        )
        .unwrap();

        let adapter = ClaudeAdapter::new(projects);
        let sessions = adapter.find_sessions();
        assert_eq!(sessions.len(), 1);
        assert_eq!(sessions[0].title, "Renamed Claude thread");
        assert_eq!(sessions[0].directory, "/work/app");
    }
}
